### PR TITLE
ESP32 gn support

### LINKED ISSUE #1510: GN build parity: esp32 platform
The GN build needs support for esp32 platform for parity with automake.

### PR DESCRIPTION
**Pull request by kedars**
#### Problem
Add gn build for ESP32

 #### Summary of Changes
fixes #1510

### CODE REVIEW

**Comment by andy31415** [comment]
Shoud this be a draft while WIP or is it ready for review?

File: src/platform/BUILD.gn [hunk 1]
```diff
@@ -261,6 +267,30 @@ if (chip_device_platform != "none" && chip_device_platform != "external") {
...
+        "ESP32/Logging.cpp",
+        "ESP32/LwIPCoreLock.cpp",
+        "ESP32/NetworkProvisioningServerImpl.h",
+        "ESP32/PlatformManagerImpl.cpp",
+        "ESP32/PlatformManagerImpl.h",
+        "ESP32/SoftwareUpdateManagerImpl.h",
+        "ESP32/SystemTimeSupport.cpp",
+        "FreeRTOS/SystemTimeSupport.cpp",
+      ]
+      #      public_configs += [ "${chip_root}/src/crypto:crypto_config" ]
```
Comment (mspang) [inline src/platform/BUILD.gn:293]: remove

File: src/system/BUILD.gn [hunk 2]
```diff
@@ -106,8 +106,10 @@ source_set("system_config_header") {
 
   public_deps = [ ":system_buildconfig" ]
 
-  if (chip_system_config_use_lwip) {
-    public_deps += [ "${chip_root}/src/lwip" ]
+  if (target_cpu != "esp32") {
```
Comment (mspang) [inline src/system/BUILD.gn:109]: target_cpu -> current_cpu

Comment (mspang) [reply src/system/BUILD.gn:109]: Can you move this condition inside src/lwip/BUILD.gn ? See e.g. third_party/mbedtls/BUILD.gn

File: src/ble/BUILD.gn [hunk 3]
```diff
@@ -40,6 +40,10 @@ buildconfig_header("ble_buildconfig") {
     defines +=
         [ "BLE_PLATFORM_CONFIG_INCLUDE=${chip_ble_platform_config_include}" ]
   }
+  if (chip_device_platform == "esp32") {
+    defines +=
+        [ "BLE_PLATFORM_CONFIG_INCLUDE=${chip_ble_platform_config_include}" ]
+  }
```
Comment (mspang) [inline src/ble/BUILD.gn:46]: Is this block intended to be added? It looks like BLE_PLATFORM_CONFIG_INCLUDE is set twice here.

File: config/esp32/toolchain/BUILD.gn [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import("//build_overrides/chip.gni")
+
+import("//build/toolchain/gcc_toolchain.gni")
+
+declare_args() {
+  ar_ext = ""
```
Comment (mspang) [inline config/esp32/toolchain/BUILD.gn:20]: If they're defined here, probably makes sense to name them something like

esp32_ar
esp32_cc
esp32_cxx

File: config/esp32/toolchain/BUILD.gn [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+gcc_toolchain("esp32") {
+  ar = ar_ext
+  cc = cc_ext
+  cxx = cxx_ext
+
+  toolchain_args = {
+    current_os = "freertos"
+    current_cpu = cpu_ext
```
Comment (mspang) [inline config/esp32/toolchain/BUILD.gn:33]: Can this be "xtensa" ? The cpu variable is typically used to determine the architecture.

File: src/crypto/BUILD.gn [hunk 6]
```diff
@@ -60,7 +60,11 @@ static_library("crypto") {
   public_configs = []
   if (chip_crypto == "mbedtls") {
     sources += [ "CHIPCryptoPALmbedTLS.cpp" ]
-    public_deps += [ "${mbedtls_root}:mbedtls" ]
+
+    # ESP32 has its own mbedTLS
+    if (target_cpu != "esp32") {
+      public_deps += [ "${mbedtls_root}:mbedtls" ]
```
Comment (mspang) [inline src/crypto/BUILD.gn:66]: Remove this and set the |mbedtls_target| argument.

File: src/ble/BUILD.gn (hunk 3 above)
Comment (kedars) [reply src/ble/BUILD.gn:42]: Ah, the upper 'if' condition had a typo, fixed.

File: config/esp32/toolchain/BUILD.gn (hunk 4 above)
Comment (kedars) [reply config/esp32/toolchain/BUILD.gn]: fixed

File: src/crypto/BUILD.gn (hunk 6 above)
Comment (kedars) [reply src/crypto/BUILD.gn]: fixed

File: config/esp32/toolchain/BUILD.gn (hunk 5 above)
Comment (kedars) [reply config/esp32/toolchain/BUILD.gn]: I've been going back and forth on that myself. Finally I thought there is nothing 'xtensa' specific here, the toolchain is esp32-specific, the SDK is esp32-specific, so might as well call it 'esp32'. Thoughts? What typically goes into generic 'cpu' specific files?

File: src/platform/BUILD.gn (hunk 1 above)
Comment (kedars) [reply src/platform/BUILD.gn]: done

**Comment by kedars** [comment]
@mspang is there a GN reference for managing custom test scripts? 
For ESP32, qemu-based GN firmware build goes through right now, but the tests are maintained as `src/inet/tests/qemu_inet_tests.sh` and executed like https://github.com/project-chip/connectedhomeip/blob/master/scripts/tests/esp32_qemu_tests.sh#L38

File: config/esp32/toolchain/BUILD.gn (hunk 5 above)
Comment (mspang) [reply config/esp32/toolchain/BUILD.gn]: The use case is writing assembly code. See e.g. 
https://github.com/project-chip/connectedhomeip/blob/master/third_party/jlink/segger_rtt/RTT/BUILD.gn#L40

**Comment by mspang** [comment]
> @mspang is there a GN reference for managing custom test scripts?
> For ESP32, qemu-based GN firmware build goes through right now, but the tests are maintained as `src/inet/tests/qemu_inet_tests.sh` and executed like https://github.com/project-chip/connectedhomeip/blob/master/scripts/tests/esp32_qemu_tests.sh#L38

No example inside CHIP yet.

Chromium on Fuchsia provides a quite sophisticated runner script generator:
https://source.chromium.org/chromium/chromium/src/+/master:build/config/fuchsia/generate_runner_scripts.gni;l=33?q=test_runner.py%20file:fuchsia&ss=chromium%2Fchromium%2Fsrc

Those scripts can run code on either real hardware or transparently spin up QEMU.

I think we can leave QEMU as a followup of this patch. Probably the existing shell scripts can be adapted; slightly longer term I think we can integrate it with the build in a way that makes "check" work. Also python should be preferred over shell to make running on Windows easier.

File: src/inet/RawEndPoint.cpp [hunk 7]
```diff
@@ -42,6 +42,9 @@
 #include <lwip/ip.h>
 #include <lwip/raw.h>
 #include <lwip/tcpip.h>
+#if CHIP_SEPARATE_CONFIG_H
```
Comment (mspang) [inline src/inet/RawEndPoint.cpp:45]: Andrei is adding this to arch.h

**Review by mspang — APPROVED** [review]

**Review by woody-apple — COMMENTED** [review]
Approving, looks good.

@kedars, can you confirm this is no longer WIP - given you pulled it out of draft?

**Review by woody-apple — APPROVED** [review]

**Comment by kedars** [comment]
> Approving, looks good.
> 
> @kedars, can you confirm this is no longer WIP - given you pulled it out of draft?

@woody-apple Everything is done, 2 notes:
* this replaces automake build on ESP32 with GN (no CMake compatibility for CHIP, I hope that's ok?)
* QEMU tests with GN may take a while. As @mspang mentioned above we could do this in a follow-on patch. Do we want to merge it as is for now? Or should I figure out if we can make both 'automake' and 'gn' build variants work for ESP32. So that gives us a breather while we take on the qemu tests?

**Comment by kedars** [comment]
> > Approving, looks good.
> > @kedars, can you confirm this is no longer WIP - given you pulled it out of draft?
> 
> @woody-apple Everything is done, 2 notes:
> 
> * this replaces automake build on ESP32 with GN (no CMake compatibility for CHIP, I hope that's ok?)
> * QEMU tests with GN may take a while. As @mspang mentioned above we could do this in a follow-on patch. Do we want to merge it as is for now? Or should I figure out if we can make both 'automake' and 'gn' build variants work for ESP32. So that gives us a breather while we take on the qemu tests?

Ok, so I made both of these builds works
* the default is the current automake. Pass the option `CHIP_BUILD_WITH_GN=y` to the build to build with GN
* also added a GitHub workflow so the GN build is triggered as well (could you please help review the workflow changes)

I will flip the build to GN default in a follow-on PR (once I figure out what the documentation changes required to the example's README)

For the 'restyler' issue, I am not able to view what it is complaining about, if it is about using `$1` without double-quotes in the `esp_echo_app.sh`, I intend it to be without double-quotes, so it nulls out

**Comment by woody-apple** [comment]
FYI to whomever merges, there's a known issue that for edits to .workflows, we may need to force merge changes in (like this one).

**Review by rwalker-apple — APPROVED** [review]

**Comment by rwalker-apple** [comment]
@saurabhst @hawk248 @jelderton ?

**Review by saurabhst — APPROVED** [review]

**Pull request merged**